core/src/com/google/inject/internal/aop/AbstractGlueGenerator.java

```diff
@@ -0,0 +1,214 @@
+    if (proxyName.startsWith("java/") && !ClassDefining.hasPackageAccess()) {
+      proxyName = '$' + proxyName; // can't define java.* glue in same package
+    }
+    return proxyName;
+  }
+
+  /** Generates the enhancer/fast-class and returns a mapping from signature to invoker. */
+  public final Function<String, BiFunction> glue(SortedMap<String, Executable> glueMap) {
+    final MethodHandle invokerTable;
+    try {
+      byte[] bytecode = generateGlue(glueMap.values());
+      Class<?> glueClass = ClassDefining.define(hostClass, bytecode);
+      invokerTable = lookupInvokerTable(glueClass);
+    } catch (Throwable e) {
+      throw new GlueException("Problem generating " + proxyName, e);
+    }
+
+    // build optimized index for these signatures and bind it to the generated invokers
+    ToIntFunction<String> signatureTable = ImmutableStringTrie.buildTrie(glueMap.keySet());
+    return bindSignaturesToInvokers(signatureTable, invokerTable);
+  }
+
+  /** Generates enhancer/fast-class bytecode for the given constructors/methods. */
+  protected abstract byte[] generateGlue(Collection<Executable> members);
+
+  /** Lookup the invoker table; this may be represented by a function or a trampoline. */
+  protected abstract MethodHandle lookupInvokerTable(Class<?> glueClass) throws Throwable;
+
+  /** Combines the signature and invoker tables into a mapping from signature to invoker. */
+  private static Function<String, BiFunction> bindSignaturesToInvokers(
+      ToIntFunction<String> signatureTable, MethodHandle invokerTable) {
+
+    // single-argument method; the table must be a function from integer index to invoker function
+    if (invokerTable.type().parameterCount() == 1) {
+      return signature -> {
+        try {
+          // pass this signature's index into the table function to retrieve the invoker
+          return (BiFunction) invokerTable.invokeExact(signatureTable.applyAsInt(signature));
+        } catch (Throwable e) {
+          throw asIfUnchecked(e);
+        }
+      };
+    }
+
+    // otherwise must be a trampoline method that takes the index and invoker arguments all at once
+    return signature -> {
+      // bind the index as soon as we have the signature...
+      int index = signatureTable.applyAsInt(signature);
+      return (instance, arguments) -> {
+        try {
+          // ...but delay calling trampoline until invocation time when we have the other arguments
+          return invokerTable.invokeExact(index, instance, (Object[]) arguments);
+        } catch (Throwable e) {
+          throw asIfUnchecked(e);
+        }
+      };
+    };
+  }
+
+  @SuppressWarnings("unchecked")
+  /** Generics trick to get compiler to treat given exception as if unchecked (as JVM does). */
+  private static <E extends Throwable> RuntimeException asIfUnchecked(Throwable e) throws E {
+    throw (E) e;
+  }
+
+  /**
+   * Generate trampoline that takes an index, along with a context object and array of argument
```

Comment:
does "and array of argument objects" imply there's a penalty for using the trampoline, in order to allocate the array & stuff the args into it?  is this a change from how FastClass or Enhancer worked?

[have we done any benchmarks on the old vs new AOP approaches?]

Reply:
For fast-invocation we are given an array of objects (which gets passed to either the fast-invoker or JDK reflection.)

Likewise for enhanced types, any intercepted methods need to convert their original arguments into an array of objects so they can be passed through an InvocationHandler and onto the method interceptor as MethodInvocation details. Methods that don't have any interceptors are not enhanced.

So the object array is a necessary detail of any solution (either having it passed in or expecting it to be provided for interception purposes.)

Re: benchmarks, I've run various tests that show performance is better using the new approach. For method interception the improvement per-invocation is small, but there's a much larger benefit when it comes to the up-front cost of interception when creating new injectors.